packages-exp/auth-exp/src/core/mfa/mfa_session.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as externs from '@firebase/auth-types-exp';
+import { debugFail } from '../util/assert';
+
+export enum MultiFactorSessionType {
+  ENROLL = 'enroll',
+  SIGN_IN = 'signin'
+}
+
+interface SerializedMultiFactorSession {
+  multiFactorSession: {
+    idToken?: string;
+    pendingCredential?: string;
+  };
+}
+
+export class MultiFactorSession implements externs.MultiFactorSession {
+  constructor(
+    readonly type: MultiFactorSessionType,
+    readonly credential: string
+  ) {}
+
+  toPlainObject(): SerializedMultiFactorSession {
+    switch (this.type) {
+      case MultiFactorSessionType.ENROLL:
+        return {
+          multiFactorSession: { idToken: this.credential }
+        };
+      case MultiFactorSessionType.SIGN_IN:
+        return {
+          multiFactorSession: { pendingCredential: this.credential }
+        };
+      default:
+        return debugFail('unexpected MultiFactorSessionType');
+    }
+  }
+
+  static fromPlainObject(
+    obj: Partial<SerializedMultiFactorSession>
+  ): MultiFactorSession | null {
+    if (obj && 'multiFactorSession' in obj && obj.multiFactorSession) {
```

Comment:
why can't you do:
```ts
if (obj?.multiFactorSession) {
  if (obj.multiFactorSession?.pendingCredential) {
    // Etc
  }
}
```

Reply:
because i'm using a Partial, this closer matches all the checks that the existing SDK does, the `in` allows the type checker to believe that the field is defined

happy to discuss more

Reply:
done